optimade/models/utils.py

```diff
@@ -1,4 +1,34 @@
-__all__ = ("CHEMICAL_SYMBOLS", "EXTRA_SYMBOLS", "ATOMIC_NUMBERS")
+import re
+
+__all__ = ("CHEMICAL_SYMBOLS", "EXTRA_SYMBOLS", "ATOMIC_NUMBERS", "SemanticVersion")
+
+
+class SemanticVersion(str):
+    """ A custom type for a semantic version, using the recommended
+    semver regexp from
+    https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string.
+
+    """
+
+    @classmethod
+    def __get_validators__(cls):
+        yield cls.validate
+
+    @classmethod
+    def __modify_schema__(cls, field_schema):
+        field_schema.update(
+            # pattern=r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$",
+            examples=["0.10.1", "1.0.0-rc.2", "1.2.3-rc.5+develop"],
+        )
+
+    def validate(cls, v: str):
+        regexp = re.compile(
+            r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$"
+        )
+        if not regexp.match(v):
+            raise ValueError(f"Unable to validate version {v} as a semver.")
+
+        return v
```

Comment:
```suggestion
class SemanticVersion(str):
    """ A custom type for a semantic version, using the recommended
    semver regexp from
    https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string.
    """

    regex = re.compile(
        r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$"
    )

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def __modify_schema__(cls, field_schema):
        field_schema.update(
            examples=["0.10.1", "1.0.0-rc.2", "1.2.3-rc.5+develop"],
        )

    @classmethod
    def validate(cls, v: str):
        if not cls.regex.match(v):
            raise ValueError(f"Unable to validate version {v} as a semver.")
        return SemanticVersion(v)

    @property
    def _match(self):
        """ The regex match """
        return self.regex.match(self)

    @property
    def major(self) -> int:
        """ The major version number 
            Only changes when incompatible changes are made
        """
        return int(self._match.group(1))

    @property
    def minor(self) -> int:
        """ The minor version number
            Changes when backwards compatible functionality is added
        """
        return int(self._match.group(2))

    @property
    def patch(self) -> int:
        """ The patch version number
            Changes when backwards compatible bugfixes
        """
        return int(self._match.group(3))

    @property
    def prerelease(self) -> str:
        """ The pre-release tag """
        return self._match.group(4)

    @property
    def build_metadata(self) -> str:
        """ The build metadata """
        return self._match.group(5)

    @property
    def base_version(self) -> str:
        """ The base version string without patch and metadata info"""
        return f"{self.major}.{self.minor}.{self.patch}"
```

Reply:
Huh, I tried it with the regex as a instance attribute but it got angry at me. Seems to work now though, guess I must have missed a @classmethod somewhere. These other additions look useful too.

Reply:
I removed the extra descriptions from the docstrings, as I think semver stands on its own as definition, plus didnt want to add caveats about e.g. pre-v1...

Reply:
Sounds good. This is a suggestion on its own.